qiskit_machine_learning/neural_networks/circuit_qnn.py

```diff
@@ -69,11 +69,14 @@ def __init__(self, circuit: QuantumCircuit,
             QiskitMachineLearningError: if `interpret` is passed without `output_shape`.
         """
 
-        # TODO: need to handle case without a quantum instance
         if isinstance(quantum_instance, (BaseBackend, Backend)):
             quantum_instance = QuantumInstance(quantum_instance)
-        self._quantum_instance = quantum_instance
-        self._sampler = CircuitSampler(quantum_instance, param_qobj=False, caching='all')
+        if quantum_instance:
+            self._quantum_instance = quantum_instance
+            self._sampler = CircuitSampler(quantum_instance, param_qobj=False, caching='all')
+        else:
+            self._quantum_instance = None
+            self._sampler = None
 
```

Comment:
This is probably done simpler as
```
        self._quantum_instance = quantum_instance
        if quantum_instance is not None:
            self._sampler = CircuitSampler(quantum_instance, param_qobj=False, caching='all')
```

Also the test should be against None explicitly and not try to treat the object as boolean. I see some original code ought to be done that way too.

I would ask now what happens in the case of None and self._sampler is not created. I imagine this ought to be done in the quantum_instance setter. Maybe it should always be done in the setter so that there is a new circuit sampler running on the instance just being provided rather than the old one running on whatever the quantum instance was before.

And again on the existing code the quantum_instance setter should have the same signature as the constructor i.e. they should be more like this https://github.com/Qiskit/qiskit-terra/blob/3b86cda6923809067588a706fda7a411e8c05264/qiskit/algorithms/variational_algorithm.py#L84-L95 which you can see logic in the setter similar to the constructor here in wrapping the backend

Reply:
I agree with @woodsp-ibm:

- There should be a new test that tests `CircuitQNN` with `quantum_instance` set to `None`.
- In the setter of `quantum_instance` we should recreate the sampler.
- Type hints should be consistent. I know, there's no type hint in the setter, but this is a mistake. Ideally, the code in the constructor and in the setter should not be duplicated.

Reply:
Would it not be better, now that the quantum instance setter takes care of creating a CircuitSampler etc, that the constructor here leveraged that code i.e used the setter directly. It might mean moving the code around a little to set self._circuit earlier so that logic around measurements will again be handled by the setter and not needed in the constructor. Better from the viewpoint that we just have one copy of the code to maintain/fix etc rather than what seems to be duplicate of code in constructor and the setter, that could be avoided.